### DISCUSSION TITLE
Add support for session-level filters

### DISCUSSION THREAD

**Discussion by Hawxy** [JasperFx/marten#2800] [Ideas]
This was mentioned in Discord few months ago. The primary purpose would be to filter entities based on additional dimensions beyond tenant, this is particularly useful if your entities are nested by some common grouping and/or you have authorization requirements around such groupings. 

As an example scenario a tenant might have most of its entities nested by location, and only users with access to those locations should have access to those entities, thus as a developer I want to ensure that all data loaded from a session takes that restriction into account.

```cs
// Pseudo example

// Our authorize middleware enforces that the user has access to the location
[Authorize("RequireLocation")]
[Route("api/{location}/[controller]")]
public class ProductController : ControllerBase
{
    string Tenant => HttpContext!.User.FindFirstValue("tenant_id")!;

    private readonly IDocumentSession _session;
    public ProductController(IDocumentStore store)
    {
        _session = store.LightweightSession(Tenant);
       // We also need to enforce that entities the user are requesting belong to this location to prevent bypass
        _session.WithFilter<ILocationEntity>(x => x.LocationId == HttpContext.GetRouteValue("location"));        
    }

    [HttpGet("{id}")]    
    public async Task<Product> Get(Guid id)
    {
       return await _session.LoadAsync<Product>(id);
    }

}
```

**Comment by mysticmind** [comment]
Moving this to discussion until we put further thoughts on this and can pull it back to issues once we are ready to look at this.
